### PR TITLE
Write about cache-predictor-based side channels

### PR DESCRIPTION
**Pull request by kbeyls**
I expect that after introducing this text, it becomes possible to then go on an start writing about transient attacks.

- - -
This change is [Reviewable](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204)

### CODE REVIEW

**Comment by kbeyls** [comment]
I had this piece of text lying around in a draft state for quite some time now.
I thought I'd put it up for review, even though I expect it may need a bit more tweaking still to get it good enough to merge.

It does seem like I need to rebase to the most recent ToT - I'll look into doing that soon.

**Comment by kbeyls** [comment]
I ended up rebasing this work onto top of main to ensure we keep on having a linear git history, not needing merges.
I hope the force push/rebasing will not confuse reviewable...

**Review by g-kouv — APPROVED** [review]
Thanks, Kristof, LGTM overall, with one suggestion!

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-:-NY2yfiY-YKZsD1WmuMB:b-8rr8gm)* status: all files reviewed, 1 unresolved discussion (waiting on @kbeyls)

___
*[`book.md` line 1763 at r1](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-NY2vh07EHPeLDA-BCn6:-NY2vh07EHPeLDA-BCn7:b667m6p) ([raw file](https://github.com/llsoftsec/llsoftsecbook/blob/3e2b82d94066d7b61df23eb85b02142eabe2dc47/book.md#L1763)):*
> ```Markdown
> instruction that needs to be executed is only known when that branch instruction
> completes. However, waiting for the branch instruction to finish before starting
> the next instruction leads to a big performance loss. Therefore, most CPUs
> ```

Perhaps it would be useful to mention here that the performance loss becomes more pronounced as pipelines get deeper and when we have more instructions in flight (in modern superscalar CPUs, https://en.wikipedia.org/wiki/Superscalar_processor).

**Review by kbeyls — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204)* status: all files reviewed, 1 unresolved discussion (waiting on @g-kouv)

___
*[`book.md` line 1763 at r1](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-NY2vh07EHPeLDA-BCn6:-NY3EitG1VTmroPdyZbL:b-kqngor) ([raw file](https://github.com/llsoftsec/llsoftsecbook/blob/3e2b82d94066d7b61df23eb85b02142eabe2dc47/book.md#L1763)):*
<details><summary><i>Previously, g-kouv (Georgia Kouveli) wrote…</i></summary><blockquote>

Perhaps it would be useful to mention here that the performance loss becomes more pronounced as pipelines get deeper and when we have more instructions in flight (in modern superscalar CPUs, https://en.wikipedia.org/wiki/Superscalar_processor).
</blockquote></details>

It is obviously true that performance loss becomes more pronounced with deeping pipelines and more instructions in flight. However, I think that covering that in this section doesn't add too much extra value. In this paragraph I'm trying to briefly explain why branch predictors exist in the first place. I think that is justified with the text as is.
I think that going deeper into it may distract from the focus of this section - which is the side channel provided by branch predictors.
So, I think it may be best, for now, to not add more information on the performance impact of branch predictors. If it turns out to be useful to go deeper on the performance impact of branch predictors, or other aspects of them, maybe that should be separated out into a separate section.
In summary, I think that for now, I'd like to leave out your suggestion - and we can add it in a later iteration if we see that extra info is needed?

**Review by g-kouv — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204)* status: all files reviewed, 1 unresolved discussion (waiting on @kbeyls)

___
*[`book.md` line 1763 at r1](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-NY2vh07EHPeLDA-BCn6:-NY3YMiCDORpHaCV8Nsh:b-9j8pxh) ([raw file](https://github.com/llsoftsec/llsoftsecbook/blob/3e2b82d94066d7b61df23eb85b02142eabe2dc47/book.md#L1763)):*
<details><summary><i>Previously, kbeyls (Kristof Beyls) wrote…</i></summary><blockquote>

It is obviously true that performance loss becomes more pronounced with deeping pipelines and more instructions in flight. However, I think that covering that in this section doesn't add too much extra value. In this paragraph I'm trying to briefly explain why branch predictors exist in the first place. I think that is justified with the text as is.
I think that going deeper into it may distract from the focus of this section - which is the side channel provided by branch predictors.
So, I think it may be best, for now, to not add more information on the performance impact of branch predictors. If it turns out to be useful to go deeper on the performance impact of branch predictors, or other aspects of them, maybe that should be separated out into a separate section.
In summary, I think that for now, I'd like to leave out your suggestion - and we can add it in a later iteration if we see that extra info is needed?
</blockquote></details>

Sure, no problem! I was just thinking that maybe the big performance loss isn't obvious for someone not already familiar with the topic, if they open the first wikipedia link and see the five stage pipeline for the first time. Perhaps we can add an appendix with some additional common terms/pointers to resources in the future (but of course we're not trying to write a microarchitecture book :)).

**Review by JLouisKaplan-Arm — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204)* status: all files reviewed, 3 unresolved discussions (waiting on @kbeyls)

___
*[`book.md` line 1792 at r1](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-NYHxaZC0-bOg7Tw3XwC:-NYHxaZC0-bOg7Tw3XwD:b-exmwak) ([raw file](https://github.com/llsoftsec/llsoftsecbook/blob/3e2b82d94066d7b61df23eb85b02142eabe2dc47/book.md#L1792)):*
> ```Markdown
>   infer whether a branch is taken or not taken in a victim process. They do so
>   by carefully making sure that a branch in the spy process uses the same branch
>   predictor entry (i.e., aliases) with the targeted branch in the victim
> ```

I think the `,` comma after "i.e." is not necessary
___
*[`book.md` line 1800 at r1](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-NYHxgT_-MZBozYLmfb2:-NYHxgT_-MZBozYLmfb3:bixl5qw) ([raw file](https://github.com/llsoftsec/llsoftsecbook/blob/3e2b82d94066d7b61df23eb85b02142eabe2dc47/book.md#L1800)):*
> ```Markdown
>   cache-based side channel attacks](#primeprobe).
> 
>   When whether a branch is taken or not depends on a bit in a secret key, this
> ```

I think "when" here is not needed

**Review by kbeyls — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204)* status: 1 of 2 files reviewed, 3 unresolved discussions (waiting on @g-kouv and @JLouisKaplan-Arm)

___
*[`book.md` line 1763 at r1](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-NY2vh07EHPeLDA-BCn6:-NYIwtve2ux_GKnPC51M:b1s6mer) ([raw file](https://github.com/llsoftsec/llsoftsecbook/blob/3e2b82d94066d7b61df23eb85b02142eabe2dc47/book.md#L1763)):*
<details><summary><i>Previously, g-kouv (Georgia Kouveli) wrote…</i></summary><blockquote>

Sure, no problem! I was just thinking that maybe the big performance loss isn't obvious for someone not already familiar with the topic, if they open the first wikipedia link and see the five stage pipeline for the first time. Perhaps we can add an appendix with some additional common terms/pointers to resources in the future (but of course we're not trying to write a microarchitecture book :)).
</blockquote></details>

Maybe we can do this with a footnote? I'll try to write one (probably sometime tomorrow).
___
*[`book.md` line 1792 at r1](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-NYHxaZC0-bOg7Tw3XwC:-NYIwl2e5tRQnK4iDok1:b-896fix) ([raw file](https://github.com/llsoftsec/llsoftsecbook/blob/3e2b82d94066d7b61df23eb85b02142eabe2dc47/book.md#L1792)):*
<details><summary><i>Previously, JLouisKaplan-Arm (Jonathan Louis Kaplan) wrote…</i></summary><blockquote>

I think the `,` comma after "i.e." is not necessary
</blockquote></details>

Done.
___
*[`book.md` line 1800 at r1](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-NYHxgT_-MZBozYLmfb2:-NYIwm-qD59ihL5lzCgQ:b-qz2s8w) ([raw file](https://github.com/llsoftsec/llsoftsecbook/blob/3e2b82d94066d7b61df23eb85b02142eabe2dc47/book.md#L1800)):*
<details><summary><i>Previously, JLouisKaplan-Arm (Jonathan Louis Kaplan) wrote…</i></summary><blockquote>

I think "when" here is not needed
</blockquote></details>

I tried phrasing this a bit differently. Do you think the new phrasing is OK?

**Review by JLouisKaplan-Arm — COMMENTED** [review]
New wording makes more sense to me, thanks Kristof!

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-:-NYMw0mdDTHLeU_wb0-F:b-p3dd26)* status: 1 of 2 files reviewed, 1 unresolved discussion (waiting on @g-kouv)

**Review by kbeyls — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @g-kouv and @JLouisKaplan-Arm)

___
*[`book.md` line 1763 at r1](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-NY2vh07EHPeLDA-BCn6:-NYMzt20D4H7fDhRtWNJ:bnns0ie) ([raw file](https://github.com/llsoftsec/llsoftsecbook/blob/3e2b82d94066d7b61df23eb85b02142eabe2dc47/book.md#L1763)):*
<details><summary><i>Previously, kbeyls (Kristof Beyls) wrote…</i></summary><blockquote>

Maybe we can do this with a footnote? I'll try to write one (probably sometime tomorrow).
</blockquote></details>

I tried to describe this in the new footnote [^branch-prediction-performance]. I struggled a bit with trying to explain this clearly. Please have a look & let my know what you think. If you'd happen to have any suggestions on rephrasing to make the footnote easier to understand, that would be very welcome!

**Review by g-kouv — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-:-NYbNCh-4iDKeUGl8Z_W:bnfp4nl)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @JLouisKaplan-Arm and @kbeyls)

___
*[`book.md` line 1787 at r5](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204#-NYbN6w_8CT-Z_b3Zv67:-NYbN6w_8CT-Z_b3Zv68:b2t8k67) ([raw file](https://github.com/llsoftsec/llsoftsecbook/blob/aaab94afbc758dbb11ee68fb4eb2612cedfb7a5a/book.md#L1787)):*
> ```Markdown
> [^branch-prediction-performance]: Over time, new CPU designs tend to support
> having more instructions in flight. [@Eyerman2009, section 4.2.3] suggests that
> branch prediction accuracy has to grow more than linear when the number of
> ```

nit: linear -> linearly?

**Review by kbeyls — COMMENTED** [review]
> Reviewed 1 of 2 files at r4, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/llsoftsec/llsoftsecbook/204)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @kbeyls)
